### PR TITLE
fix: cache slot to header data while checking BABE equivocation

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- Refactor `verifyBlockEquivocation` by using a slot state, similar to how substrate works, https://github.com/paritytech/substrate/blob/c14ff62f0a2bca89f02f8e1cb7c1347f1d2a0baa/client/consensus/babe/src/lib.rs#L1221

## Tests

```sh
go test -timeout 10m -run ^Test_checkEquivocation$ github.com/ChainSafe/gossamer/dot/state --tags=integration -v
```

## Issues

- Closes https://github.com/ChainSafe/gossamer/issues/3317

## Primary Reviewer

@kishansagathiya

### CODE REVIEW

File: dot/state/service.go [hunk 1]
```diff
@@ -34,6 +34,7 @@ type Service struct {
 	Transaction *TransactionState
 	Epoch       *EpochState
 	Grandpa     *GrandpaState
+	Slot        *SlotState
```
Comment (jimjbrettj) [inline dot/state/service.go:37]: Should slot state be a part of babe state?

File: dot/state/slot_test.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	header3 := createHeader(t, 2) // @ slot 4
+	header4 := createHeader(t, 3) // @ slot MAX_SLOT_CAPACITY + 4
+	header5 := createHeader(t, 4) // @ slot MAX_SLOT_CAPACITY + 4
+	header6 := createHeader(t, 3) // @ slot 4
+
+	slotState := NewSlotState(inMemoryDB)
+
+	// It's ok to sign same headers.
+	equivProf, err := slotState.CheckEquivocation(2, 2, header1, aliceAuthorityID)
+	require.Nil(t, err)
```
Comment (jimjbrettj) [inline dot/state/slot_test.go:84]: nit: require.NoError() here and other usages

**Review by jimjbrettj — COMMENTED** [review]
Looks great, just a few comments

File: dot/state/service.go (hunk 1 above)
Comment (EclesioMeloJunior) [reply dot/state/service.go:37]: the slot state is a dependency for the BABE verifier, which is part of the BABE service but I don't think it should be part of the BABE state as in substrate they have a module only for slots (`client/consensus/slots`)

**Review by dimartiro — APPROVED** [review]
Great work Eclésio! 👏  LGTM

**Review by jimjbrettj — APPROVED** [review]
great work!

**Pull request merged**